lib/engine/step/g_18_co/moving_bid_auction.rb

```diff
@@ -0,0 +1,215 @@
+        end
+
+        def max_bid(player, company)
+          player.cash - committed_cash(player) + current_bid_amount(player, company)
+        end
+
+        def moveable_bids(player, company)
+          @bids.map do |cmp, company_bids|
+            next if cmp == company
+
+            player_bids = company_bids.select { |bid| bid.entity == player }
+            next if player_bids.empty?
+
+            [cmp, player_bids]
+          end.compact.to_h
+        end
+
+        protected
+
+        # every company is always up for auction
+        def can_auction?(_company)
+          true
+        end
+
+        def all_passed!
+          @bids.each do |company, bids|
+            resolve_bids_for_company(company, bids)
+          end
+        end
+
+        def resolve_bids_for_company(company, bids)
+          return if bids.empty?
+
+          # Companies without bids can be bought be corporations later
+          # Unsure how that will be accomplished at this time
+          high_bid = highest_bid(company)
+          buy_company(high_bid.entity, company, high_bid.price)
+        end
+
+        def buy_company(player, company, price)
+          company.owner = player
+          player.companies << company
+          player.spend(price, @game.bank) if price.positive?
+          @companies.delete(company)
+
+          @log << "#{player.name} wins the auction for #{company.name} "\
+                  "with #{@bids[company].size > 1 ? 'a' : 'the only'} "\
+                  "bid of #{@game.format_currency(price)}"
+
+          company.abilities(:shares) do |ability|
+            ability.shares.each do |share|
+              if share.president
+                @round.companies_pending_par << company
+              else
+                @game.share_pool.buy_shares(player, share, exchange: :free)
+              end
+            end
+          end
+        end
+
+        def accept_bid(bid)
+          price = bid.price
+          company = bid.company
+          player = bid.entity
+          @bids.delete(company)
+          buy_company(player, company, price)
+        end
+
+        def add_bid(bid)
+          company = bid.company || bid.corporation
+          entity = bid.entity
+          price = bid.price
+          min = min_bid(company)
+
+          @game.game_error("Minimum bid is #{@game.format_currency(min)} for #{company.name}") if price < min
+          if @game.class::MUST_BID_INCREMENT_MULTIPLE && ((price - min) % min_increment).nonzero?
+            @game.game_error("Must increase bid by a multiple of #{@game.format_currency(min_increment)}")
+          end
+          if price > max_bid(entity, company)
+            @game.game_error("Cannot afford bid. Maximum possible bid is
+              #{@game.format_currency(max_bid(entity, company))}")
+          end
+
+          bids = @bids[company]
+          highest_player_bid = bids.sort(&:price).reverse.find { |b| b.entity == entity }
```

Comment:
can you bid twice? why do you need to sort?

Reply:
Yes, you can have multiple bids on a single company. For example, you could have a low bid on Company B, get outbid and then move a bid from Company A to Company B.

Reply:
I'm going to standardize the method for finding the highest player bid using max as we identified earlier